### ISSUE TITLE
Problem when interfacing with GtkD in GGPlotD.drawToSurface

### ISSUE THREAD

**Issue by ljubobratovicrelja** [BlackEdder/ggplotd#21]
Hi there!

I'm making an image plotting utility for the [library](https://github.com/ljubobratovicrelja/dcv) I'm working on, and I'd like to extend that utility so it can also draw plots from GGPlotD context. [Here is the gist](https://gist.github.com/ljubobratovicrelja/98a84e00d09206583b68cecfa48a8a1e) to the short preview of such utility. (Only important parts are left in the code, read on for problem explanation please)

I was planning to [wrap](https://gist.github.com/ljubobratovicrelja/98a84e00d09206583b68cecfa48a8a1e#file-plot-d-L16-L47) the `GGPlotD.drawToSurface` call inside the utility's API, but I've got the following problem - In this utility I'm using cairo.ImageSurface from the GtkD library, but the input parameter to `drawToSurface` is actually from the cairoD library. Even if `ggplotd:ggplotd-gtk` configuration is used, cairo modules are used from cairoD, and not from GtkD library, which makes interfacing with GtkD code bit problematic. 

Dirty solution I've got in given gist, was to create the `cairoD ImageSurface` for the `drawToSurface`, and later to copy it's results to my original `GtkD ImageSurface`. Would it be possible to use cairo modules from GtkD in the ggplotd code if ggplotd-gtk configuration is used?

Also, there was another problem with `drawToSurface` method, more precisely with `ref cairo.Surface` as input. I'm still quite new to D, so I don't know why (but I suppose you do) this code:

``` d
Surface surface = new ImageSurface(Format.CAIRO_FORMAT_RGB24, w, h);
surface = gg.drawToSurface(surface, w, h);
```

... gives me this error (DMD64 D Compiler v2.071.0):

```
Error: function ggplotd.ggplotd.GGPlotD.drawToSurface (ref Surface surface, int width, int height) const is not callable using argument types (Surface, int, int) const
```

I've buy-passed this by removing the `ref`. May I ask, why was ref needed? Method produces valid results without it. Am I missing something here?

**Comment by BlackEdder** [comment]
With regards to the gtkd/cairod overlap. Both are based on the same c code and allow access to the underlying c pointers. It is safe to cast from the cairod c pointer to the gtkd c pointer. I do this in the gtk.d file:
https://github.com/BlackEdder/ggplotd/blob/e22e24272bbc4292df1123cf63e031d9c5467062/source/ggplotd/gtk.d#L49

It would probably be a good idea to add a drawToSurface method that takes a gtkd surface and turns it into cairod. It might take me some time to do that (PRs would be very welcome).

That error seems weird. Are you passing it a cairod or a gtkd surface?

You are right that the ref is not really needed (surface is a class, so always passed as ref). Added the ref to make it extra clear that the passed surface will be changed. I don't really understand why removing the ref should make any difference. Can you produce a minimal code example that gives the error you are encountering?

**Comment by ljubobratovicrelja** [comment]
Thanks for a swift reply!

> Both are based on the same c code and allow access to the underlying c pointers. It is safe to cast from the cairod c pointer to the gtkd c pointer.

Awesome, didn't know this. Thanks!

> It would probably be a good idea to add a drawToSurface method that takes a gtkd surface and turns it into cairod. It might take me some time to do that (PRs would be very welcome).

If casting is legal between cairod and gtkd pointers, this would be trivial to do, right? I'll try to put up a PR.

> That error seems weird. Are you passing it a cairod or a gtkd surface?

cairoD - [here](https://gist.github.com/ljubobratovicrelja/98a84e00d09206583b68cecfa48a8a1e#file-plot-d-L2-L4) you can see I'm aliasing the GtkD module differently.

> Can you produce a minimal code example that gives the error you are encountering?

Without the `ref` removal, [this](https://gist.github.com/ljubobratovicrelja/98a84e00d09206583b68cecfa48a8a1e#file-plot-d-L21-L22) example produces the error. Given code there compiles only if `ref` is removed. But I'll prepare the easy-to-compile dummy project that produces the error as soon as I can.

**Comment by BlackEdder** [comment]
> If casting is legal between cairod and gtkd pointers, this would be trivial to do, right? I'll try to put up a PR.

Yes it would be pretty straightforward (around 5 lines of code). Make sure to wrap it in the ggplotdGTK version block though, so that gtk remains an optional dependency.

> But I'll prepare the easy-to-compile dummy project that produces the error as soon as I can.

I'd appreciate that. With the newest dub, you can also put the dub dependency at top of your file and make it easily compilable:https://forum.dlang.org/post/nj65kc$ecn$1@digitalmars.com

**Comment by ljubobratovicrelja** [comment]
> Make sure to wrap it in the ggplotdGTK version block though, so that gtk remains an optional dependency.

Deal!

> But I'll prepare the easy-to-compile dummy project that produces the error as soon as I can.

I'm terribly sorry for taking your time on this - In my code, I've made a cast from ImageSurface to Surface when passing it to the drawToSurface, so the compile error was therefore in place:

``` d
cairo.ImageSurface surface = new cairo.ImageSurface(cairo.Format.CAIRO_FORMAT_RGB24, w, h);
surface = cast(cairo.ImageSurface)gg.drawToSurface(cast(cairo.Surface)surface, w, h);

```

So in contrast, following code compiles as expected:

``` d
cairo.Surface surface = new cairo.ImageSurface(cairo.Format.CAIRO_FORMAT_RGB24, w, h);
surface = gg.drawToSurface(surface, w, h);
```

This is embarrassing, and again sorry for the false alarm. But thanks for the support anyways!

**Comment by BlackEdder** [comment]
> I'm terribly sorry for taking your time on this - In my code, I've made a cast from ImageSurface to Surface when passing it to the drawToSurface, so the compile error was therefore in place:

No worries. Happy to hear that you solved it!

**Comment by BlackEdder** [comment]
Fixed by #22
